### PR TITLE
chore(rootfs): nginx 1.18.0, openssl 1.1.1g

### PR DESCRIPTION
**Pull request by felixbuenemann**
changes to software versions:
* nginx 1.18.0:
  * removed deprecated ipv6 build flag
  * use optimized cloudflare zlib fork
* openssl 1.1.1g with patches:
  * support ChaChaPoly-1305-Draft cipher used by older Androids
  * support cipher suite priorization for TLS 1.3 in Nginx
  * support equal preference ciphers groups (eg. ChaCha vs. AES, based on BoringSSL)
  * build with weak cipher support to keep 3DES support
* modsecurity 3.0.4
* mod_security_nginx 3.3 + fixes
* owasp_mod_security_crs latest 3.3/dev ruleset
* ngx_http_geoip2_module 3.3

new features available in router:
* tls 1.3 support
* tls 1.3 early data (0-rtt), including mandatory Early-Data request header for apps
* chacha-poly and chacha-poly-draft ciphers
* equal preferemce cipher groups
* advanced cipher suite syntax like `@SECLEVEL=n` (validation fixes)

changes to router configuration:
* tls 1.3 enabled by default
* tls 1.3 early data enabled by default for GET, HEAD, OPTIONS requests
* equal preference groups for aes128-gcm and chacha-poly in default cipher suite

### CODE REVIEW

**Comment by felixbuenemann** [comment]
I've also pushed an image `felixbuenemann/router:git-4bec8bd` to Docker Hub for easy testing, it can be used as a drop-in for existing clusters by changing the image in the deis.router deployment.

File: README.md [hunk 1]
```diff
@@ -260,8 +260,8 @@ _Note that Kubernetes annotation maps are all of Go type `map[string]string`.  A
...
...256|TLS_CHACHA20_POLY1305_SHA256]:TLS_AES_256_GCM_SHA384:[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305|ECDHE-ECDSA-CHACHA20-POLY1305-OLD]:[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305|ECDHE-RSA-CHACHA20-POLY1305-OLD]:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA"` | nginx `ssl_ciphers`.  The default ciphers are taken from the intermediate compatibility section in the [Mozilla Wiki on Security/Server Side TLS](https://wiki.mozilla.org/Security/Server_Side_TLS). If the value is set to the empty string, OpenSSL's default ciphers are used.  In _all_ cases, server side cipher preferences (order matters) are used, but equal preference groups can be used to relax that. |
```
Comment (Cryptophobia) [inline README.md:264]: I believe the formatting of this file is broken @felixbuenemann ? When I preview this file in GitHub, I see that the formatting is broken. I think it has to do with the `|` characters. Maybe they need to be escaped?

File: README.md [hunk 2]
```diff
@@ -270,6 +270,7 @@ _Note that Kubernetes annotation maps are all of Go type `map[string]string`.  A
...
 | <a name="ssl-hsts-include-sub-domains"></a>deis-router | deployment | [router.deis.io/nginx.ssl.hsts.includeSubDomains](#ssl-hsts-include-sub-domains) | `"false"` | Whether to enforce HSTS for subsequent requests to all subdomains of the original request. |
 | <a name="ssl-hsts-preload"></a>deis-router | deployment | [router.deis.io/nginx.ssl.hsts.preload](#ssl-hsts-preload) | `"false"` | Whether to allow the domain to be included in the HSTS preload list. |
+| <a name="ssl-early-data-methods"></a>deis-router | deployment | [router.deis.io/nginx.ssl.earlyDataMethods](#ssl-early-data-methods) | `"GET|HEAD|OPTIONS"` | enables nginx `ssl_early_data` (TLS 1.3 0-RTT) for the listes HTTP methods (set to `""` to disable, valid methods: `"GET|HEAD|POST|PUT|DELETE|PATCH|OPTIONS"`). Unsafe or non-idempotent methods should be avoided, to prevent replay attacks. The header `Early-Data: 1` is forwarded to apps, when Early Data is used and they can reply with HTTP status 425 to block it, causing the client to retry without Early-Data. Requires "TLSv1.3" in `"protocols"` to work.|
```
Comment (Cryptophobia) [inline README.md:273]: Same thing here. The `|` character is breaking the formatting.

File: model/model.go [hunk 3]
```diff
@@ -211,32 +211,35 @@ func newCertificate(cert string, key string) *Certificate {
 // SSLConfig represents SSL-related configuration options.
 type SSLConfig struct {
 	Enforce           bool        `key:"enforce" constraint:"(?i)^(true|false)$"`
-	Protocols         string      `key:"protocols" constraint:"^((SSLv2|SSLv3|TLSv1|TLSv1\\.1|TLSv1\\.2)\\s*)+$"`
-	Ciphers           string      `key:"ciphers" constraint:"^(!?[A-Z][A-Z\\d\\+-]+:?)*$"`
+	Protocols         string      `key:"protocols" constraint:"^((SSLv[2-3]|TLSv1(?:\\.[1-3])?)\\s*)+$"`
+	Ciphers           string      `key:"ciphers" constraint:"^((\\b[\\w.!+-]+\\b)+(:?@(STRENGTH|SECLEVEL=[0-5]))?(:([!+-]\\b)?|$))*(((\\b[\\w.+-]+\\b)+|(\\[(\\b[\\w.|+-]+\\b)+\\]))(:|$))*$"`
```
Comment (Cryptophobia) [inline model/model.go:215]: Can you give me some examples on how this regex has changed and how the Ciphers need to annotated now?

File: model/model.go [hunk 4]
```diff
@@ -211,32 +211,35 @@ func newCertificate(cert string, key string) *Certificate {
 // SSLConfig represents SSL-related configuration options.
 type SSLConfig struct {
 	Enforce           bool        `key:"enforce" constraint:"(?i)^(true|false)$"`
-	Protocols         string      `key:"protocols" constraint:"^((SSLv2|SSLv3|TLSv1|TLSv1\\.1|TLSv1\\.2)\\s*)+$"`
-	Ciphers           string      `key:"ciphers" constraint:"^(!?[A-Z][A-Z\\d\\+-]+:?)*$"`
+	Protocols         string      `key:"protocols" constraint:"^((SSLv[2-3]|TLSv1(?:\\.[1-3])?)\\s*)+$"`
```
Comment (Cryptophobia) [inline model/model.go:214]: Good change here :100:

File: model/model.go [hunk 5]
```diff
@@ -211,32 +211,35 @@ func newCertificate(cert string, key string) *Certificate {
...
 	HSTSConfig        *HSTSConfig `key:"hsts"`
+	EarlyDataMethods  string      `key:"earlyDataMethods" constraint:"^((GET|HEAD|POST|PUT|DELETE|PATCH|OPTIONS)(\\|\\b|$))*$"`
 	DHParam           string
 }
 
 func newSSLConfig() *SSLConfig {
 	return &SSLConfig{
 		Enforce:   false,
-		Protocols: "TLSv1 TLSv1.1 TLSv1.2",
+		Protocols: "TLSv1 TLSv1.1 TLSv1.2 TLSv1.3",
```
Comment (Cryptophobia) [inline model/model.go:228]: Nice!

File: model/model.go [hunk 6]
```diff
@@ -211,32 +211,35 @@ func newCertificate(cert string, key string) *Certificate {
...
+		// Source: https://wiki.mozilla.org/Security/Server_Side_TLS (old backward compatibility)
+		Ciphers:           "[TLS_AES_128_GCM_SHA256|TLS_CHACHA20_POLY1305_SHA256]:TLS_AES_256_GCM_SHA384:[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305|ECDHE-ECDSA-CHACHA20-POLY1305-OLD]:[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305|ECDHE-RSA-CHACHA20-POLY1305-OLD]:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA",
```
Comment (Cryptophobia) [inline model/model.go:237]: I guess this answers the questions I have from above. This would be the default setting for the router if nothing is provided? :+1:

File: model/model_validation_test.go [hunk 7]
```diff
@@ -300,15 +300,15 @@ func TestInvalidSSLProtocols(t *testing.T) {
 }
 
 func TestValidSSLProtocols(t *testing.T) {
-	testValidValues(t, newTestSSLConfig, "Protocols", "protocols", []string{"SSLv3", "TLSv1", "TLSv1 TLSv1.1"})
+	testValidValues(t, newTestSSLConfig, "Protocols", "protocols", []string{"SSLv3", "TLSv1", "TLSv1 TLSv1.1", "TLSv1.2 TLSv1.3"})
 }
 
 func TestInvalidSSLCiphers(t *testing.T) {
-	testInvalidValues(t, newTestSSLConfig, "Ciphers", "ciphers", []string{"0", "-1", "foobar"})
+	testInvalidValues(t, newTestSSLConfig, "Ciphers", "ciphers", []string{"!DSS", "@STRENGTH", "!TLS_AES_128_GCM_SHA256", "[3DES]!DSS", "[TLSv1]@STRENGTH", "[TLS13]:@STRENGTH"})
```
Comment (Cryptophobia) [inline model/model_validation_test.go:307]: Improved this test. Good.

File: nginx/config.go [hunk 8]
```diff
@@ -186,6 +194,7 @@ http {
 		ssl_protocols {{ $sslConfig.Protocols }};
 		{{ if ne $sslConfig.Ciphers "" }}ssl_ciphers {{ $sslConfig.Ciphers }};{{ end }}
 		ssl_prefer_server_ciphers on;
+		ssl_early_data {{ if ne $sslConfig.EarlyDataMethods "" }}on{{ else }}off{{ end }};
```
Comment (Cryptophobia) [inline nginx/config.go:197]: So `ssl_early_data` is turned off if no methods are set here? Is that the behaviour we want as default. Seems like ssl_early_data makes everything more secure in TLSv1.3?

File: rootfs/Dockerfile [hunk 9]
```diff
@@ -14,8 +14,8 @@ RUN set -x && \
         libpcre3-dev \
         libmaxminddb-dev \
         libfuzzy-dev && \
-    export MOD_SECURITY_VERSION=3.0.3 BUILD_PATH=$PWD PREFIX=/usr/local && \
-    get_src 8aa1300105d8cc23315a5e54421192bc617a66246ad004bd89e67c232208d0f4 \
+    export MOD_SECURITY_VERSION=3.0.4 BUILD_PATH=$PWD PREFIX=/usr/local && \
```
Comment (Cryptophobia) [inline rootfs/Dockerfile:17]: Update to 3.0.4 modsecurity. :+1:

File: rootfs/Dockerfile [hunk 10]
```diff
@@ -42,39 +69,55 @@ RUN adduser --system \
...
+    cd "$BUILD_PATH/zlib-$CLOUDFLARE_ZLIB_VERSION" && \
+    make -f Makefile.in distclean && \
     cd "$BUILD_PATH/nginx-$NGINX_VERSION" && \
+    patch -p1 -i "$BUILD_PATH/nginx_auto_using_PRIORITIZE_CHACHA.patch" && \
     ./configure \
       --prefix="$PREFIX" \
       --pid-path=/tmp/nginx.pid \
       --with-debug \
       --with-pcre-jit \
-      --with-ipv6 \
```
Comment (Cryptophobia) [inline rootfs/Dockerfile:77]: I wasn't sure if ipv6 support is really needed, good removal. :) Maybe some day in the future.

File: README.md (hunk 1 above)
Comment (felixbuenemann) [reply README.md:264]: Yes, that will be confused as table columns. I'll look into it.

File: model/model.go (hunk 3 above)
Comment (felixbuenemann) [reply model/model.go:215]: The previous regex was very limited and only matched a small subset of what can be expressed as cipher suites, but existing annotations will continue to work unchanged.

For example it would reject EECDH+AESGCM:EECDH+CHACHA20:+aRSA because it required ciphers to start with an uppercase letter, optionally preceded by an exclamation mark, which isn't even valid as the first term. So eg 3DES would be valid, but rejected, but !DSS would be allowed, but is invalid on its own (but DHE+AES:!DSS would be OK).

You have to take a look at the openssl ciphers manpage/docs to understand all the rules and I tried to encode as many as reasonably possible in a short-ish regex. 

In addition to the openssl ciphers rules it also encodes additional rules that apply when equal preference groups are used, because some more advanced stuff can only be done before the first preference group, as is denoted in the comments of the patch ported from boringssl. 

The patch also allows to list TLS 1.3 ciphers like TLS_AES_128_GCM_SHA256, so the rules had to be extended for that.

In short this regex was a pain in the ass to write and took me hours to get right, especially since go doesn't support lookahead and lookbehind assertions in regular expressions…

File: model/model.go (hunk 6 above)
Comment (felixbuenemann) [reply model/model.go:237]: Yes, it will pretty much have the same intent as the previous ciphers, but extended for TLS 1.3 and allowing client preference between AES128-GCM and ChaCha20. 

Essentially the equal preference groups allow to give the client some choice in choosing the best cipher while still applying server side ordering of cipher preference. 

This is important, since many mobile client lack AES hardware acceleration and ChaCha20 is much cheaper that AESGCM in software, but desktop clients usually have AES-NI, so fot the AESGCM is cheaper than ChaCha20.

File: nginx/config.go (hunk 8 above)
Comment (felixbuenemann) [reply nginx/config.go:197]: Yes, that can be used to disable it. The default is "GET|HEAD|OPTIONS" which are the HTTP methods considered idempotent and safe. 

Early-Data actually makes things less secure, but speeds up initial connections from 3 to 2 roundtrips, making them as fast as session resumptions.

Early-Data allows to send the http request like GET / as part of the SSL handshake, but since the handshake is not completed yet, it can be used for replay attacks. 

If someone captures the tcp packets of a connection handshake with early data, those packets cannot be decoded, but can be replayed.

This is the reason that only safe and idempotent HTTP methods are allowed by default, since replaying them should be harmless.

Of course if application developers ignore HTTP conventions and use GET requests to delete records or initiate payments, they are in trouble. 

This is why the router adds the standard "Early-Data: 1" request header and apps can reply with the "425 Too Early" HTTP status code, which would cause the client to retry the request after the handshake has completed.

File: rootfs/Dockerfile (hunk 10 above)
Comment (felixbuenemann) [reply rootfs/Dockerfile:77]: It isn't, it is always on, nginx just removed the configure flag for it.

File: model/model.go (hunk 3 above)
Comment (Cryptophobia) [reply model/model.go:215]: Well I'm glad you spent time perfecting this regex and everyone will appreciate this new router update soon.

File: model/model.go (hunk 6 above)
Comment (Cryptophobia) [reply model/model.go:237]: Very nice and this is knowledge that I did not have before because I never really had to read up on it. I know that newer CPUs have hardware acceleration for AES in most new architectures, but didn't know that ChaCha20 is preferred for mobile because of this. 

> As Adam Langley described, ChaCha20-Poly1305 is three times faster than AES-128-GCM on mobile devices. Spending less time on decryption means faster page rendering and better battery life. ... On desktop computers with hardware AES support, AES-128-GCM is still the faster choice.Feb 23, 2015

File: nginx/config.go (hunk 8 above)
Comment (Cryptophobia) [reply nginx/config.go:197]: Very cool! Thank you for explaining this. Sounds like Early-Data is only for optimization purposes added to TLSv1.3 . Although less-secure, I can see why developers may want it enabled for: 

> "GET|HEAD|OPTIONS" which are the HTTP methods considered idempotent and safe. 

If app developers are using GET request to delete records or initiate payments, then they have more serious problems to worry about than this nginx router. :smile:

File: rootfs/Dockerfile (hunk 10 above)
Comment (Cryptophobia) [reply rootfs/Dockerfile:77]: Okay, sounds like this is just nginx source code and makefile configuration changes. :+1:

**Comment by Cryptophobia** [comment]
@felixbuenemann , if you fix the README markdown tables for the ciphers, we are good to go and I can merge this in before making a new minor release. :)

File: README.md (hunk 1 above)
Comment (felixbuenemann) [reply README.md]: Fixed in 94e77aeb8f9f646ebfb2f26bd72aabc2f8156308.

File: model/model.go (hunk 6 above)
Comment (felixbuenemann) [reply model/model.go:237]: Fixed in 94e77aeb8f9f646ebfb2f26bd72aabc2f8156308.

**Comment by felixbuenemann** [comment]
README is fixed in 94e77aeb8f9f646ebfb2f26bd72aabc2f8156308, so I think this is good to go from my side.
